pkg/healthcheck/healthcheck.go

```diff
@@ -251,9 +255,10 @@ type CheckResult struct {
 type CheckObserver func(*CheckResult)
 
 type category struct {
-	id       CategoryID
-	checkers []checker
-	enabled  bool
+	id        CategoryID
+	checkers  []checker
+	enabled   bool
```

Comment:
This feels a bit redundant. What are the benefits of introducing `shouldRun`? Is it possible to determine the HA condition on the caller side? AFAICT, it has been that the clients (`install`, `check`, `public_api`) decide which checks to include, after determining the check categories based on some conditions.

Reply:
So I guided myself only by the check.go use of this class but the way I see it is that the decisions of which checks to include is based solely on the command line option. To make that depend on for example whether the installation is HA or not, we need to fetch the config map in advanced and potentially fail or default to false if we cannot get the config map. But this is already part of the check sequence tself and in fact comes later than some other checks. Namely: 


```
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

kubernetes-version
------------------
√ is running the minimum Kubernetes API version
√ is running the minimum kubectl version

linkerd-config
--------------
√ control plane Namespace exists
√ control plane ClusterRoles exist
√ control plane ClusterRoleBindings exist
√ control plane ServiceAccounts exist
√ control plane CustomResourceDefinitions exist
√ control plane MutatingWebhookConfigurations exist
√ control plane ValidatingWebhookConfigurations exist
√ control plane PodSecurityPolicies exist

linkerd-existence
-----------------
√ 'linkerd-config' config map exists.  <----- this here
``` 

So requiring to have the config map upfront in order to understand what checks we need to include, while some of the checks checking whether this config map exists at all renders the whole thing a bit leaky in my mind. Feel free to disagree and we can always change it :)

Reply:
this isn't the first time I've wanted to conditionally skip a check depending on the current state.  what do you think about having a special error value called skip?  if a check wishes to be skipped, it can return this value.  the check runner can check the error, and skip over any checks that have returned skip.

Reply:
> So requiring to have the config map upfront in order to understand what checks we need to include, while some of the checks checking whether this config map exists at all renders the whole thing a bit leaky in my mind. 

I'm not sure that's an issue though, because the CLI and some controllers already knows how to look up the configmap. I'd assume this HA check happens after the `linkerd-existence` check referred above.
```
⚡  grep -rniI fetchlinkerdconfigmap *                                         
cli/cmd/install.go:896: _, _, err = healthcheck.FetchLinkerdConfigMap(kubeAPI, controlPlaneNamespace)
cli/cmd/upgrade.go:214: _, configs, err := healthcheck.FetchLinkerdConfigMap(k, controlPlaneNamespace)
cli/cmd/profile.go:108:                 _, configs, err := healthcheck.FetchLinkerdConfigMap(k8sAPI, controlPlaneNamespace)
cli/cmd/install-sp.go:30:                       _, configs, err := healthcheck.FetchLinkerdConfigMap(k8sAPI, controlPlaneNamespace)
controller/heartbeat/heartbeat.go:25:   cm, _, err := healthcheck.FetchLinkerdConfigMap(kubeAPI, controlPlaneNamespace)
web/main.go:80: cm, _, err := healthcheck.FetchLinkerdConfigMap(k8sAPI, *controllerNamespace)
```

Reply:
@ihcsim if we go that route, we can also clean up some awkward coupling where the linkerdConfig is set by one of the checks, imposing an implicit ordering constraint.  If we're going to look up the config in the check.go we should set it on the healthcheck from the beginning.

Reply:
I agree with @ihcsim's suggestion of loading to configmap when we're building the list of checks instead of during the linkerd-config check.  This lets include the LinkerdHAChecks category only if the install is HA.

Reply:
Alright, this  makes sense +1 for the cleanup of checks fetching and setting the config map!

Reply:
With this refactor, we no longer need the `shouldRun` field, right?

Reply:
@zaharidichev is this correct?

Reply:
this is correct, I will update the code to reflect that

Reply:
@adleong @ihcsim Sorry for the back and forth on this one, but taking another look at it, I have some concerns. We can surely fetch the linkerd config map from the get go which pretty much entails the following: 
1. Initialize the kube config API
2. Query the config map
3. Parse it and determine whether we are in HA or not. 

Now that all makes sense but in doing so we get ahead of ourselves in terms of checks. Because the checks as they go are something like: 
- can initialize the client - tries to initialize the kubeApi ( and sets the kubeApi var in the HealthChecker struct)
- can query the Kubernetes API - checks whether the k8s cluster cna actually be queried
- linkerd-config' config map exists - fetches and sets the config map, etc reporting an error if it cannot do so. 

So if we fetch the config map from the get go we are effectivelly bypassing these couple of checks? What happens if we get an error? do we fail right away with an output format that is different that what a failed check would print?  Or do we fail silently so we let the actual checks run and report the problems in the usual format. In the latter case we are doing the same thing twice... 

So all in all it does not seem to me that moving the config map fetching and as a consequence the kube Api intialization out of the checks and thereby remove the checks dependency results in the akwardness being reduced. After all these two things are checks in themselves... Not sure what your thoughts on this are.

Reply:
Yup, that sure is awkward.

I think the simplest thing to do is probably to have a special error struct called `skipped` which checks can return when they want to be skipped.  The run checks code would need to handle this error properly and not print a line for checks that return this value.  This ends up being very similar to having a `shouldRun` but in a more backwards compatible way and keeps all the logic for a check in one place.

Reply:
@adleong I swtiched to a custom error type. I agree that looks a lot neater. Feel free to take another look:)